ably/realtime_conn.go

```diff
@@ -0,0 +1,367 @@
+package ably
+
+import (
+	"errors"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/ably/ably-go/ably/proto"
+
+	"github.com/ably/ably-go/Godeps/_workspace/src/code.google.com/p/go.net/websocket"
+	"github.com/ably/ably-go/Godeps/_workspace/src/gopkg.in/vmihailenco/msgpack.v2"
+)
+
+type msgerr struct {
+	msg proto.ProtocolMessage
+	err error
+}
+
+// Conn represents a single connection RealtimeClient instantiates for
+// communication with Ably servers.
+type Conn struct {
+	id        string
+	key       string
+	serial    int64
+	msgSerial int64
+	err       error
+	conn      MsgConn
+	connCh    chan *msgerr
+	msgCh     chan *proto.ProtocolMessage
+	opts      *ClientOptions
+	state     *stateEmitter
+	stateCh   chan State
+	pending   pendingEmitter
+	queue     *msgQueue
+}
+
+func newConn(opts *ClientOptions) (*Conn, error) {
+	c := &Conn{
+		opts:   opts,
+		connCh: make(chan *msgerr),
+		msgCh:  make(chan *proto.ProtocolMessage),
+		state:  newStateEmitter(StateConn, StateConnInitialized, ""),
+	}
+	c.queue = newMsgQueue(c)
+	if opts.Listener != nil {
+		c.On(opts.Listener)
+	}
+	if !opts.NoConnect {
+		if err := c.Connect(); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+func (c *Conn) dial(proto string, u *url.URL) (MsgConn, error) {
+	if c.opts.Dial != nil {
+		return c.opts.Dial(proto, u)
+	}
+	return dialWebsocket(proto, u)
+}
+
+func booltext(b ...bool) []string {
+	ok := true
+	for _, b := range b {
+		ok = ok && b
+	}
+	if ok {
+		return []string{"true"}
+	}
+	return []string{"false"}
+}
+
+// Connect is used to connect to Ably servers manually, when the client owning
+// the connection was created with NoConnect option. The connect request is
+// being processed on a separate goroutine.
+//
+// In order to receive its result register a channel with the On method; upon
+// succsufful connection channel will receive StatConnConnected event.
```

Comment:
What is the use case where I would explicitly call `Connect()` but not want to know the result straight away? It seems to me like this should return once the connection attempt has been made.